core/src/main/java/pt/webdetails/cda/push/WebsocketJsonQueryEndpoint.java

```diff
@@ -0,0 +1,225 @@
+      final QueryOptions queryOptions = core.getQueryOptions( parameters );
+      final TableExporter tableExporter = cdaEngine.getExporter( queryOptions );
+
+      StreamingDataservicesDataAccess dataAccess = (StreamingDataservicesDataAccess) cdaSettings.getDataAccess( queryOptions.getDataAccessId() );
+
+      this.consumer = PublishSubject.create();
+
+      consumer.toFlowable( BackpressureStrategy.LATEST ).
+        onBackpressureBuffer( 1, () -> { }, BackpressureOverflowStrategy.DROP_OLDEST ).
+        map( RowMetaToTableModel.getConverter()::apply ).
+        subscribe( tableModel -> {
+          try ( OutputStream outputStream = new ByteArrayOutputStream() ) {
+            tableExporter.export( outputStream, tableModel );
+            outboundMessageConsumer.accept( outputStream.toString() );
+          } catch ( Exception e ) {
+            logger.error( "Error converting a stream push table model into a json output", e );
+          }
+        } );
+
+      disposableConsumer = new DisposableObserver<>( consumer );
+
+      dataAccess.doPushStreamQuery( queryOptions, disposableConsumer );
+
+    } catch ( Exception e ) {
+      logger.error( "Error processing JSON query message.", e );
+      throw new RuntimeException( e );
+    }
+  }
+
+  @Override public void onClose() {
+    if ( this.consumer != null ) {
+      this.consumer.onComplete();
+    }
+    if ( this.disposableConsumer != null && !this.disposableConsumer.isDisposed() ) {
+      this.disposableConsumer.dispose();
+    }
+  }
+
+  private static class DisposableObserver<T> implements Observer<T>, Disposable {
+
+    private Observer<T> actual;
+    private Disposable toDispose;
+
+    public DisposableObserver( Observer<T> subject ) {
+      this.actual = subject;
+    }
+
+    @Override
+    public void dispose() {
+      if ( toDispose != null ) {
+        toDispose.dispose();
+      }
+    }
+
+    @Override
+    public boolean isDisposed() {
+      return toDispose != null && toDispose.isDisposed();
+    }
+
+    @Override
+    public void onSubscribe( Disposable d ) {
+      toDispose = d;
+      actual.onSubscribe( d );
+    }
+
+    @Override
+    public void onNext( T t ) {
+      actual.onNext( t );
+    }
+
+    @Override
+    public void onError( Throwable e ) {
+      actual.onError( e );
+    }
+
+    @Override
+    public void onComplete() {
+      actual.onComplete();
+    }
+  }
+
+  @Override
+  public List<String> getSubProtocols() {
+    return Collections.singletonList( ACCEPTED_SUB_PROTOCOL );
+  }
+
+  /**
+   * Gets a {@link MultivaluedMap} from the message received.
+   * @param message a JSON message with parameters.
+   * @return a multi value map of values obtained from the JSON message received as parameter.
+   * @throws JSONException Is thrown when there is a problem converting the message into a JSON object.
+   */
+  private MultivaluedMap<String, String> getParameters( String message ) throws JSONException {
```

Comment:
These kind of methods, it would be great if they could be publicly available and kept on core, as they provide standard utility functionality.